main_test.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeterminePackagePrefix(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "thriftrw-main-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	genDir, err := filepath.Abs("gen")
+	require.NoError(t, err)
+
+	tests := []struct {
+		around func(func())
+		dir    string
+		result string
+		errMsg string
+	}{
+		{
+			around: func(f func()) {
+				gopath := os.Getenv("GOPATH")
+				require.NoError(t, os.Unsetenv("GOPATH"))
+				defer os.Setenv("GOPATH", gopath)
+
+				f()
+			},
+			dir:    genDir,
+			errMsg: "$GOPATH is not set",
+		},
+		{
+			dir:    tmpDir,
+			errMsg: "not inside $GOPATH",
+		},
+		{
+			dir:    genDir,
+			result: "github.com/thriftrw/thriftrw-go/gen",
+		},
+	}
+
+	for _, tt := range tests {
+		around := tt.around
+		if around == nil {
+			around = func(f func()) { f() }
+		}
+
+		around(func() {
+			result, err := determinePackagePrefix(tt.dir)
+
+			if tt.result != "" {
+				if assert.NoError(t, err) {
+					assert.Equal(t, tt.result, result)
+				}
+			}
+
+			if tt.errMsg != "" {
+				if assert.Error(t, err, "determinePackagePrefix(%q) should fail", tt.dir) {
+					assert.Contains(t, err.Error(), tt.errMsg)
+				}
+			}
+		})
+	}
+}
+
```

Comment:
can we add tests for `findCommonAncestor` since it's kind of complicated (yay magic)